src/server/accept.rs

```diff
@@ -132,19 +130,20 @@ where
                     Ok(accepted) => {
                         // Create a new HTTP/2 connection.
                         #[cfg(feature = "http2")]
-                        let mut connection = conn::http2::Builder::new(TokioExecutor::new())
-                            .timer(TokioTimer::new())
-                            .serve_connection(IoStream::new(accepted), service);
+                        let mut connection = Box::pin(
+                            conn::http2::Builder::new(TokioExecutor::new())
+                                .timer(TokioTimer::new())
+                                .serve_connection(TokioIo::new(accepted), service),
+                        );
 
                         // Create a new HTTP/1.1 connection.
                         #[cfg(all(feature = "http1", not(feature = "http2")))]
-                        let mut connection = conn::http1::Builder::new()
-                            .timer(TokioTimer::new())
-                            .serve_connection(IoStream::new(accepted), service)
-                            .with_upgrades();
-
-                        // Pin the connection on the stack so it can be polled.
-                        let mut connection = Pin::new(&mut connection);
+                        let mut connection = Box::pin(
```

Comment:
I've seen a noticeable improvement with regards to predictable memory usage after making this change.

Reply:
Ended up changing my mind back. If we're going to allocate once in the connection task before hitting an await point, we should use it for the socket. That way, we know that it's memory address will not change under any circumstances.